Extensions/Guardrail+Settings.swift

```diff
@@ -79,59 +79,70 @@ public extension Guardrail where Value == HKQuantity {
         recommendedBounds: 16...399,
         unit: HKUnit.milligramsPerDeciliter.unitDivided(by: .internationalUnit())
     )
-
+ 
     static let carbRatio = Guardrail(
-        absoluteBounds: 1...150,
-        recommendedBounds: 3.0.nextUp...28.0.nextDown,
+        absoluteBounds: 2...150,
+        recommendedBounds: 4...28,
         unit: .gramsPerUnit
     )
 
     static func basalRate(supportedBasalRates: [Double]) -> Guardrail {
-        let recommendedLowerBound = supportedBasalRates.first == 0
-            ? supportedBasalRates.dropFirst().first!
-            : supportedBasalRates.first!
+        let scheduledBasalRateAbsoluteRange = 0.05...30.0
+        let allowedBasalRates = supportedBasalRates.filter { scheduledBasalRateAbsoluteRange.contains($0) }
         return Guardrail(
-            absoluteBounds: supportedBasalRates.first!...supportedBasalRates.last!,
-            recommendedBounds: recommendedLowerBound...supportedBasalRates.last!,
+            absoluteBounds: allowedBasalRates.first!...allowedBasalRates.last!,
+            recommendedBounds: allowedBasalRates.first!...allowedBasalRates.last!,
             unit: .internationalUnitsPerHour
         )
     }
 
-    static var recommendedMaximumScheduledBasalScaleFactor: Double {
-        return 6
-    }
-
     static func maximumBasalRate(
         supportedBasalRates: [Double],
         scheduledBasalRange: ClosedRange<Double>?,
+        lowestCarbRatio: Double?,
         maximumBasalRatePrecision decimalPlaces: Int = 3
     ) -> Guardrail {
-        let minimumSupportedBasalRate = supportedBasalRates.first!
-        let recommendedLowerBound = minimumSupportedBasalRate == 0 ? supportedBasalRates.dropFirst().first! : minimumSupportedBasalRate
+        
+        let maximumUpperBound = 70.0 / (lowestCarbRatio ?? carbRatio.absoluteBounds.lowerBound.doubleValue(for: .gramsPerUnit))
+        let absoluteUpperBound = maximumUpperBound.matchingOrTruncatedValue(from: supportedBasalRates, withinDecimalPlaces: decimalPlaces)
+
+        let recommendedHighScheduledBasalScaleFactor = 6.4
+        let recommendedLowScheduledBasalScaleFactor = 2.1
+
+        let recommendedLowerBound: Double
         let recommendedUpperBound: Double
-        if let maximumScheduledBasalRate = scheduledBasalRange?.upperBound {
-            let scaledMaximumScheduledBasalRate = (recommendedMaximumScheduledBasalScaleFactor * maximumScheduledBasalRate).matchingOrTruncatedValue(from: supportedBasalRates, withinDecimalPlaces: decimalPlaces)
-            recommendedUpperBound = maximumScheduledBasalRate == 0
+        if let highestScheduledBasalRate = scheduledBasalRange?.upperBound {
+            recommendedLowerBound = (recommendedLowScheduledBasalScaleFactor * highestScheduledBasalRate).matchingOrTruncatedValue(from: supportedBasalRates, withinDecimalPlaces: decimalPlaces)
+            recommendedUpperBound = highestScheduledBasalRate == 0
```

Comment:
[nit] I don't think you need this check here. It can just be:
```
recommendedUpperBound = (recommendedHighScheduledBasalScaleFactor * highestScheduledBasalRate).matchingOrTruncatedValue(from: supportedBasalRates, withinDecimalPlaces: decimalPlaces)
```
It will result in the same value without the special condition.

Reply:
Yeah, I had to stare at that, too. This was code I didn't change, so thanks for confirming my suspicions!